File: CodeGenDotNet 159059-1372 (a)/CSharp/Tools/CodeGenerationSupport/XSLTSupport.cs
// ====================================================================
//  Copyright ©2004, Kathleen Dollard, All Rights Reserved.
// ====================================================================
//   I'm distributing this code so you'll be able to use it to see code
//   generation in action and I hope it will be useful and you'll enjoy 
//   using it. This code is provided "AS IS" without warranty, either 
//   expressed or implied, including implied warranties of merchantability 
//   and/or fitness for a particular purpose. 
// ====================================================================
//  Summary: Provides support for running XSLT templates
//  Refactor: There are too many translation methods

using System;

namespace KADGen.CodeGenerationSupport
{
	public struct Param
	{
		public string Name;
		public object Value;
		public Param( string Name, object value )
		{
			this.Name = Name;
			this.Value = value;
		}
		public Param( string Name )
		{
			this.Name = Name;
			this.Value = null;
		}
	}
	public struct ExtObject
	{
		public string NameSpaceURI;
		public object value;
		public ExtObject( string NameSpaceURI, object value )
		{
			this.NameSpaceURI = NameSpaceURI;
			this.value = value;
		}
		public ExtObject( string NameSpaceURI )
		{
			this.NameSpaceURI = NameSpaceURI;
			this.value = null;
		}
	}
	public class XSLTSupport
	{
		
		#region Class level declarations - empty
		#endregion

		#region constructors - empty
		#endregion

		#region public Methods and Properties
		public static void GenerateViaXSLT( string xsltFileName, string MetaDataFileName, string outputFile, string outputType, ExtObject[] extObjects, params Param[] @params )
		{
			System.Xml.XmlDocument xmlDoc = new System.Xml.XmlDocument();
			xmlDoc.Load( MetaDataFileName );
			KADGen.LibraryInterop.Singletons.XMLDoc = xmlDoc;
			KADGen.LibraryInterop.Singletons.NsMgr = KADGen.Utility.xmlHelpers.BuildNamespacesManagerForDoc( xmlDoc );
			GenerateViaXSLT( xsltFileName, xmlDoc, outputFile, outputType, extObjects, @params );
		}


		public static void GenerateViaXSLT( string xsltFileName, System.Xml.XmlDocument xmlMetaData, string outputFile, string outputType, ExtObject[] extObjects, params Param[] @params )
		{
			System.IO.FileStream stream = null;
			try
			{
				stream = new System.IO.FileStream( outputFile, System.IO.FileMode.Create );
				GenerateXSLTToStream( xsltFileName, xmlMetaData, stream, outputType, extObjects, @params );
			}
			finally
			{
				stream.Close();
			}

		}

		public static System.IO.Stream GenerateViaXSLT( string xsltFileName, System.Xml.XmlDocument xmlMetaData, string outputType, ExtObject[] extObjects, params Param[] @params )
		{
			System.IO.MemoryStream stream;
			stream = new System.IO.MemoryStream();
			GenerateXSLTToStream( xsltFileName, xmlMetaData, stream, outputType, extObjects, @params );
			stream.Flush();
			stream.Seek( 0, System.IO.SeekOrigin.Begin );
			return stream;
		}

		public static System.IO.Stream GenerateViaXSLT( System.Xml.Xsl.XslTransform xslTransform, System.Xml.XPath.XPathNavigator xNavMetaData, string outputType, ExtObject[] extObjects, params Param[] @params )
		{
			System.IO.MemoryStream stream;
			stream = new System.IO.MemoryStream();
			GenerateXSLTToStream( xslTransform, xNavMetaData, stream, outputType, extObjects, @params );
			stream.Flush();
			stream.Seek( 0, System.IO.SeekOrigin.Begin );
			return stream;
		}

		public static System.IO.Stream GenerateViaXSLT( string xsltFileName, string MetaDataFileName, string outputType, ExtObject[] extObjects, params Param[] @params )
		{
			System.Xml.XmlDocument xmlDoc = new System.Xml.XmlDocument();
			xmlDoc.Load( MetaDataFileName );
			KADGen.LibraryInterop.Singletons.XMLDoc = xmlDoc;
			KADGen.LibraryInterop.Singletons.NsMgr = KADGen.Utility.xmlHelpers.BuildNamespacesManagerForDoc( xmlDoc );
			return GenerateViaXSLT( xsltFileName, xmlDoc, outputType, extObjects, @params );
		}

		public static void GenerateXMLViaXSLT( string xsltFileName, System.Xml.XmlDocument xmlMetaData, string outputFile, string outputType, ExtObject[] extObjects, params Param[] @params )
		{
			System.Xml.Xsl.XslTransform xslt = new System.Xml.Xsl.XslTransform();
			System.Xml.XPath.XPathNavigator xNav;
			System.Xml.XmlTextWriter XMLWriter = null;
			System.Xml.Xsl.XsltArgumentList args = new System.Xml.Xsl.XsltArgumentList();

			try
			{
				if( xmlMetaData == null )
				{
					xmlMetaData = new System.Xml.XmlDocument();
				}

				foreach( Param param in @params )
				{
					args.AddParam( param.Name, "", param.Value );
				}

				if( extObjects == null )
				{
					// No problem, just skip
				}
				else
				{
					foreach( ExtObject extObject in extObjects )
					{
						System.Reflection.ConstructorInfo constructorInfo = ((System.Type)(extObject.value)).GetConstructor( null );
						object obj = constructorInfo.Invoke( null );
						args.AddExtensionObject( extObject.NameSpaceURI, obj );
					}
				}
				AddStandardExtension( args );

				xNav = xmlMetaData.CreateNavigator();
				XMLWriter = new System.Xml.XmlTextWriter( outputFile, System.Text.Encoding.UTF8 );

				xslt.Load( xsltFileName );
				xslt.Transform( xNav, args, XMLWriter, null );

			}
			catch( System.Exception ex )
			{
				System.Diagnostics.Debug.WriteLine( ex );
				throw;

			}
			finally
			{
				if( XMLWriter != null )
				{
					XMLWriter.Flush();
					XMLWriter.Close();
				}
			}

		}

		public static Param[] GetXSLTParams( string xsltFileName, string basePath )
		{
			System.IO.Stream xsltStream = null;
			System.Xml.XmlDocument xmlInput = new System.Xml.XmlDocument();
			System.IO.MemoryStream stream = new System.IO.MemoryStream();
			string[] paramNames;
			Param[] @params = null;
			//System.Xml.XmlNode attr;
			//string value;

			try
			{
				xsltStream = Utility.Tools.GetStreamFromStringResource( typeof(XSLTSupport), "RetrieveParams.xslt" );
				xmlInput.Load( xsltFileName );
				GenerateXSLTToStream( xsltStream, xmlInput, stream, "", null );
				stream.Seek( 0, System.IO.SeekOrigin.Begin );
				System.IO.StreamReader reader = new System.IO.StreamReader( stream );
				paramNames = reader.ReadToEnd().Split( '|' );
				if( paramNames.GetLength(0 ) == 1 && paramNames[0].Length == 0 )
				{
					// there are no parameters
				}
				else
				{
					@params = new Param[paramNames.GetLength( 0 )];
					for( int i = 0; i <= paramNames.GetUpperBound(0 ); i++ )
					{
						if( paramNames[i].Trim().Length > 0 )
						{
							@params[i] = new Param( paramNames[i], "");
						}
					}
				}
			}
			finally
			{
				try
				{
					stream.Close();
				}
				catch
				{
				}
				try
				{
					xsltStream.Close();
				}
				catch
				{
				}
			}
			return @params;
		}



		public static void FillParams( Param[] @params, System.Xml.XPath.XPathNavigator nav )
		{
			for( int i=0; i<@params.Length; i++ )
			{
				@params[i].Value = nav.GetAttribute( @params[i].Name, "" );
			}
		}

		#endregion

		#region protected and internal Methods and Properties - empty
		#endregion

		#region private Methods and Properties
		private static void GenerateXSLTToStream( string xsltFileName, System.Xml.XmlDocument xmlMetaData, System.IO.Stream stream, string outputType, ExtObject[] extObjects, params Param[] @params )
		{
			System.Xml.Xsl.XslTransform xslt = new System.Xml.Xsl.XslTransform();
			System.Xml.XPath.XPathNavigator xNav;
			System.IO.StreamWriter streamWriter = new System.IO.StreamWriter( stream );
			System.Xml.Xsl.XsltArgumentList args = new System.Xml.Xsl.XsltArgumentList();

			try
			{
				if( xmlMetaData == null )
				{
					xmlMetaData = new System.Xml.XmlDocument();
				}

				xNav = xmlMetaData.CreateNavigator();

				xslt.Load( xsltFileName );
				GenerateXSLTToStream( xslt, xNav, stream, outputType, extObjects, @params );

			}
			catch( System.Exception ex )
			{
				System.Diagnostics.Debug.WriteLine( ex );
				throw;

			}
			finally
			{
				if( streamWriter != null )
				{
					streamWriter.Flush();
				}
			}
		}

		private static void GenerateXSLTToStream( System.IO.Stream xsltStream, System.Xml.XmlDocument xmlMetaData, System.IO.Stream stream, string outputType, ExtObject[] extObjects, params Param[] @params )
		{
			System.Xml.Xsl.XslTransform xslt = new System.Xml.Xsl.XslTransform();
			System.Xml.XPath.XPathNavigator xNav;
			System.IO.StreamWriter streamWriter = new System.IO.StreamWriter( stream );
			System.Xml.Xsl.XsltArgumentList args = new System.Xml.Xsl.XsltArgumentList();
			System.Xml.XmlTextReader xmlReader;

			try
			{
				if( xmlMetaData == null )
				{
					xmlMetaData = new System.Xml.XmlDocument();
				}

				xNav = xmlMetaData.CreateNavigator();

				xmlReader = new System.Xml.XmlTextReader( xsltStream );
				xslt.Load( xmlReader, null, new System.Security.Policy.Evidence() );
				GenerateXSLTToStream( xslt, xNav, stream, outputType, extObjects, @params );

			}
			catch( System.Exception ex )
			{
				System.Diagnostics.Debug.WriteLine( ex );
				throw ex;

			}
			finally
			{
				if( streamWriter != null )
				{
					streamWriter.Flush();
				}
			}
		}

		private static void GenerateXSLTToStream( System.Xml.Xsl.XslTransform xsltTransform, System.Xml.XPath.XPathNavigator xNavMetaData, System.IO.Stream stream, string outputType, ExtObject[] extObjects, params Param[] @params )
		{
			System.IO.StreamWriter streamWriter = null;
			System.Xml.XmlTextWriter xmlWriter = null;
			System.Xml.Xsl.XsltArgumentList args = new System.Xml.Xsl.XsltArgumentList();

			try
			{
				if( @params != null )
				{
					foreach( Param param in @params )
					{
						args.AddParam( param.Name, "", param.Value );
					}
				}

				if( extObjects == null )
				{
					// No problem, just skip
				}
				else
				{
					foreach( ExtObject extObject in extObjects )
					{
						System.Reflection.ConstructorInfo constructorInfo = ((System.Type)(extObject.value)).GetConstructor( Type.EmptyTypes );
						object obj = constructorInfo.Invoke( null );
						args.AddExtensionObject( extObject.NameSpaceURI, obj );
					}
				}
				AddStandardExtension( args );

				streamWriter = new System.IO.StreamWriter( stream );
				xsltTransform.Transform( xNavMetaData, args, streamWriter, null );

			}
			catch( System.Exception ex )
			{
				System.Diagnostics.Debug.WriteLine( ex );
				throw;

			}
			finally
			{
				if( streamWriter != null )
				{
					streamWriter.Flush();
				}
				if( xmlWriter != null )
				{
					xmlWriter.Flush();
				}
			}
		}


		private static void AddStandardExtension( System.Xml.Xsl.XsltArgumentList args )
		{
			args.AddExtensionObject( "http://kadgen.com/StandardNETSupport.xsd", new XSLTSupport.StandardSupport() );
		}
		#endregion

		private class StandardSupport
		{
			
			private int mIndent = 2;
			private int mIndentSize = 2;

			public void SetIndentSize( int value )
			{
				mIndentSize = value;
			}

			public int GetIndent(   )
			{
				return mIndent;
			}

			public void SetIndent( int value )
			{
				mIndent = value;
			}

			public void Indent(   )
			{
				mIndent += 1;
			}

			public void Outdent(   )
			{
				mIndent -= 1;
			}

			public string InsertIndent(   )
			{
				System.Text.StringBuilder sb = new System.Text.StringBuilder();
				int repeatCount = mIndentSize * mIndent;
				if( repeatCount > 0 )
				{
					sb.Append( ' ', mIndentSize * mIndent );
				}
				return sb.ToString();
			}

			public string InsertNLIndent()
			{
				return '\n' + InsertIndent();
			}
		}
	}
}

File: CodeGenDotNet 159059-1372 (a)/VBCode/Tools/CodeGenerationSupport/XSLTSupport.vb
' ====================================================================
'  Copyright ©2004, Kathleen Dollard, All Rights Reserved.
' ====================================================================
'   I'm distributing this code so you'll be able to use it to see code
'   generation in action and I hope it will be useful and you'll enjoy 
'   using it. This code is provided "AS IS" without warranty, either 
'   expressed or implied, including implied warranties of merchantability 
'   and/or fitness for a particular purpose. 
' ====================================================================
'  Summary: Provides support for running XSLT templates
'  Refactor: There are too many translation methods

Option Strict On
Option Explicit On

Imports System
Imports System.Diagnostics

'! Class Summary: 

Public Structure Param
   Public Name As String
   Public Value As Object
   Public Sub New(ByVal Name As String, ByVal Value As Object)
      Me.Name = Name
      Me.Value = Value
   End Sub
   Public Sub New(ByVal Name As String)
      Me.Name = Name
   End Sub
End Structure

Public Structure ExtObject
   Public NameSpaceURI As String
   Public Value As Object
   Public Sub New(ByVal NameSpaceURI As String, ByVal Value As Object)
      Me.NameSpaceURI = NameSpaceURI
      Me.Value = Value
   End Sub
   Public Sub New(ByVal NameSpaceURI As String)
      Me.NameSpaceURI = NameSpaceURI
   End Sub
End Structure

Public Class XSLTSupport

#Region "Class level declarations - empty"
#End Region

#Region "Constructors - empty"
#End Region

#Region "Public Methods and Properties"
   Public Shared Sub GenerateViaXSLT( _
               ByVal xsltFileName As String, _
               ByVal MetaDataFileName As String, _
               ByVal outputFile As String, _
               ByVal outputType As String, _
               ByVal extObjects() As ExtObject, _
               ByVal ParamArray params() As Param)
      Dim xmlDoc As New Xml.XmlDocument
      xmlDoc.Load(MetaDataFileName)
      KADGen.LibraryInterop.Singletons.XMLDoc = xmlDoc
      KADGen.LibraryInterop.Singletons.NsMgr = KADGen.Utility.xmlHelpers.BuildNamespacesManagerForDoc(xmlDoc)
      GenerateViaXSLT(xsltFileName, xmlDoc, outputFile, outputType, extObjects, params)
   End Sub


   Public Shared Sub GenerateViaXSLT( _
               ByVal xsltFileName As String, _
               ByVal xmlMetaData As Xml.XmlDocument, _
               ByVal outputFile As String, _
               ByVal outputType As String, _
                ByVal extObjects() As ExtObject, _
               ByVal ParamArray params() As Param)
      Dim stream As IO.FileStream
      Try
         stream = New IO.FileStream(outputFile, IO.FileMode.Create)
         GenerateXSLTToStream(xsltFileName, xmlMetaData, stream, outputType, extObjects, params)
      Finally
         stream.Close()
      End Try

   End Sub

   Public Shared Function GenerateViaXSLT( _
               ByVal xsltFileName As String, _
               ByVal xmlMetaData As Xml.XmlDocument, _
               ByVal outputType As String, _
               ByVal extObjects() As ExtObject, _
               ByVal ParamArray params() As Param) _
               As IO.Stream
      Dim stream As IO.MemoryStream
      stream = New IO.MemoryStream
      GenerateXSLTToStream(xsltFileName, xmlMetaData, stream, outputType, extObjects, params)
      stream.Flush()
      stream.Seek(0, IO.SeekOrigin.Begin)
      Return stream
   End Function

   Public Shared Function GenerateViaXSLT( _
               ByVal xslTransform As Xml.Xsl.XslTransform, _
               ByVal xNavMetaData As Xml.XPath.XPathNavigator, _
               ByVal outputType As String, _
               ByVal extObjects() As ExtObject, _
               ByVal ParamArray params() As Param) _
               As IO.Stream
      Dim stream As IO.MemoryStream
      stream = New IO.MemoryStream
      GenerateXSLTToStream(xslTransform, xNavMetaData, stream, outputType, extObjects, params)
      stream.Flush()
      stream.Seek(0, IO.SeekOrigin.Begin)
      Return stream
   End Function

   Public Shared Function GenerateViaXSLT( _
               ByVal xsltFileName As String, _
               ByVal MetaDataFileName As String, _
               ByVal outputType As String, _
               ByVal extObjects() As ExtObject, _
               ByVal ParamArray params() As Param) _
               As IO.Stream
      Dim xmlDoc As New Xml.XmlDocument
      xmlDoc.Load(MetaDataFileName)
      KADGen.LibraryInterop.Singletons.XMLDoc = xmlDoc
      KADGen.LibraryInterop.Singletons.NsMgr = KADGen.Utility.xmlHelpers.BuildNamespacesManagerForDoc(xmlDoc)
      Return GenerateViaXSLT(xsltFileName, xmlDoc, outputType, extObjects, params)
   End Function

   Public Shared Sub GenerateXMLViaXSLT( _
               ByVal xsltFileName As String, _
               ByVal xmlMetaData As Xml.XmlDocument, _
               ByVal outputFile As String, _
               ByVal outputType As String, _
               ByVal extObjects() As ExtObject, _
               ByVal ParamArray params() As Param)
      Dim xslt As New Xml.Xsl.XslTransform
      Dim xNav As Xml.XPath.XPathNavigator
      Dim XMLWriter As Xml.XmlTextWriter
      Dim args As New Xml.Xsl.XsltArgumentList
      Dim param As param

      Try
         If xmlMetaData Is Nothing Then
            xmlMetaData = New Xml.XmlDocument
         End If

         For Each param In params
            args.AddParam(param.Name, "", param.Value)
         Next

         If extObjects Is Nothing Then
            ' No problem, just skip
         Else
            For Each extObject As extObject In extObjects
               Dim constructorInfo As Reflection.ConstructorInfo = CType(extObject.Value, System.Type).GetConstructor(Nothing)
               Dim obj As Object = constructorInfo.Invoke(Nothing)
               args.AddExtensionObject(extObject.NameSpaceURI, obj)
            Next
         End If
         AddStandardExtension(args)

         xNav = xmlMetaData.CreateNavigator()
         XMLWriter = New Xml.XmlTextWriter(outputFile, Text.Encoding.UTF8)

         xslt.Load(xsltFileName)
         xslt.Transform(xNav, args, XMLWriter, Nothing)

      Catch ex As System.Exception
         Debug.WriteLine(ex)
         Throw

      Finally
         If Not XMLWriter Is Nothing Then
            XMLWriter.Flush()
            XMLWriter.Close()
         End If
      End Try

   End Sub

   Public Shared Function GetXSLTParams( _
            ByVal xsltFileName As String, _
            ByVal basePath As String) _
            As Param()
      Dim xsltStream As IO.Stream
      Dim xmlInput As New Xml.XmlDocument
      Dim stream As New IO.MemoryStream
      Dim paramNames() As String
      Dim params() As Param
      Dim attr As Xml.XmlNode
      Dim value As String

      Try
         xsltStream = Utility.Tools.GetStreamFromStringResource( _
                     GetType(XSLTSupport), "RetrieveParams.xslt")
         xmlInput.Load(xsltFileName)
         GenerateXSLTToStream(xsltStream, xmlInput, stream, "", Nothing)
         stream.Seek(0, IO.SeekOrigin.Begin)
         Dim reader As New IO.StreamReader(stream)
         paramNames = reader.ReadToEnd.Split("|"c)
         If paramNames.GetLength(0) = 1 AndAlso paramNames(0).Length = 0 Then
            ' there are no parameters
         Else
            ReDim params(paramNames.GetUpperBound(0))
            For i As Int32 = 0 To paramNames.GetUpperBound(0)
               If paramNames(i).Trim.Length > 0 Then
                  params(i) = New Param(paramNames(i), "")
               End If
            Next
         End If
      Finally
         Try
            stream.Close()
         Catch
         End Try
         Try
            xsltStream.Close()
         Catch
         End Try
      End Try
      Return params
   End Function



   Public Shared Sub FillParams( _
                  ByVal params() As Param, _
                  ByVal nav As Xml.XPath.XPathNavigator)
      For Each param As param In params
         param.Value = nav.GetAttribute(param.Name, "")
      Next
   End Sub

#End Region

#Region "Protected and Friend Methods and Properties - empty"
#End Region

#Region "Private Methods and Properties"
   Private Shared Sub GenerateXSLTToStream( _
               ByVal xsltFileName As String, _
               ByVal xmlMetaData As Xml.XmlDocument, _
               ByVal stream As IO.Stream, _
               ByVal outputType As String, _
               ByVal extObjects() As ExtObject, _
               ByVal ParamArray params() As Param)
      Dim xslt As New Xml.Xsl.XslTransform
      Dim xNav As Xml.XPath.XPathNavigator
      Dim streamWriter As New IO.StreamWriter(stream)
      Dim args As New Xml.Xsl.XsltArgumentList
      Dim param As param

      Try
         If xmlMetaData Is Nothing Then
            xmlMetaData = New Xml.XmlDocument
         End If

         xNav = xmlMetaData.CreateNavigator()

         xslt.Load(xsltFileName)
         GenerateXSLTToStream(xslt, xNav, stream, outputType, extObjects, params)

      Catch ex As System.Exception
         Debug.WriteLine(ex)
         Throw

      Finally
         If Not streamWriter Is Nothing Then
            streamWriter.Flush()
         End If
      End Try
   End Sub

   Private Shared Sub GenerateXSLTToStream( _
            ByVal xsltStream As IO.Stream, _
            ByVal xmlMetaData As Xml.XmlDocument, _
            ByVal stream As IO.Stream, _
               ByVal outputType As String, _
             ByVal extObjects() As ExtObject, _
            ByVal ParamArray params() As Param)
      Dim xslt As New Xml.Xsl.XslTransform
      Dim xNav As Xml.XPath.XPathNavigator
      Dim streamWriter As New IO.StreamWriter(stream)
      Dim args As New Xml.Xsl.XsltArgumentList
      Dim param As param
      Dim xmlReader As Xml.XmlTextReader

      Try
         If xmlMetaData Is Nothing Then
            xmlMetaData = New Xml.XmlDocument
         End If

         xNav = xmlMetaData.CreateNavigator()

         xmlReader = New Xml.XmlTextReader(xsltStream)
         xslt.Load(xmlReader, Nothing, New Security.Policy.Evidence)
         GenerateXSLTToStream(xslt, xNav, stream, outputType, extObjects, params)

      Catch ex As System.Exception
         Debug.WriteLine(ex)
         Throw

      Finally
         If Not streamWriter Is Nothing Then
            streamWriter.Flush()
         End If
      End Try
   End Sub

   'Private Shared Sub GenerateXSLTToStream( _
   '            ByVal xsltFileName As String, _
   '            ByVal xmlMetaData As Xml.XmlDocument, _
   '            ByVal stream As IO.Stream, _
   '            ByVal ParamArray params() As XSLTParam)
   '   Dim xslt As New Xml.Xsl.XslTransform
   '   Dim xNav As Xml.XPath.XPathNavigator
   '   Dim streamWriter As New IO.StreamWriter(stream)
   '   Dim args As New Xml.Xsl.XsltArgumentList
   '   Dim param As XSLTParam

   '   Try
   '      If xmlMetaData Is Nothing Then
   '         xmlMetaData = New Xml.XmlDocument
   '      End If

   '      For Each param In params
   '         args.AddParam(param.Name, "", param.Value)
   '      Next

   '      xNav = xmlMetaData.CreateNavigator()

   '      xslt.Load(xsltFileName)
   '      xslt.Transform(xNav, args, streamWriter, Nothing)

   '   Catch ex As System.Exception
   '      Debug.WriteLine(ex)
   '      Throw

   '   Finally
   '      If Not streamWriter Is Nothing Then
   '         streamWriter.Flush()
   '      End If
   '   End Try
   'End Sub

   Private Shared Sub GenerateXSLTToStream( _
               ByVal xsltTransform As Xml.Xsl.XslTransform, _
               ByVal xNavMetaData As Xml.XPath.XPathNavigator, _
               ByVal stream As IO.Stream, _
               ByVal outputType As String, _
               ByVal extObjects() As ExtObject, _
               ByVal ParamArray params() As Param)
      Dim streamWriter As IO.StreamWriter
      Dim xmlWriter As Xml.XmlTextWriter
      Dim args As New Xml.Xsl.XsltArgumentList
      Dim param As param

      Try
         If Not params Is Nothing Then
            For Each param In params
               args.AddParam(param.Name, "", param.Value)
            Next
         End If

         If extObjects Is Nothing Then
            ' No problem, just skip
         Else
            For Each extObject As extObject In extObjects
               Dim constructorInfo As Reflection.ConstructorInfo = CType(extObject.Value, System.Type).GetConstructor(Type.EmptyTypes)
               Dim obj As Object = constructorInfo.Invoke(Nothing)
               args.AddExtensionObject(extObject.NameSpaceURI, obj)
            Next
         End If
         AddStandardExtension(args)

         'If outputType.Trim.ToUpper = "XML" Then
         '   xmlWriter = New Xml.XmlTextWriter(stream, Text.Encoding.UTF8)
         '   xsltTransform.Transform(xNavMetaData, args, xmlWriter, Nothing)
         'Else
            streamWriter = New IO.StreamWriter(stream)
            xsltTransform.Transform(xNavMetaData, args, streamWriter, Nothing)
         'End If

      Catch ex As System.Exception
         Debug.WriteLine(ex)
         Throw

      Finally
         If Not streamWriter Is Nothing Then
            streamWriter.Flush()
         End If
         If Not xmlWriter Is Nothing Then
            xmlWriter.Flush()
         End If
      End Try
   End Sub


   Private Shared Sub AddStandardExtension(ByVal args As Xml.Xsl.XsltArgumentList)
      args.AddExtensionObject("http://kadgen.com/StandardNETSupport.xsd", _
            New XSLTSupport.StandardSupport)

   End Sub

#End Region

   Private Class StandardSupport
      Private mIndent As Int32 = 2
      Private mIndentSize As Int32 = 2

      Public Sub SetIndentSize(ByVal value As Int32)
         mIndentSize = value
      End Sub

      Public Function GetIndent() As Int32
         Return mIndent
      End Function

      Public Sub SetIndent(ByVal Value As Int32)
         mIndent = Value
      End Sub

      Public Sub Indent()
         mIndent += 1
      End Sub

      Public Sub Outdent()
         mIndent -= 1
      End Sub

      Public Function InsertIndent() As String
         Dim sb As New Text.StringBuilder
         Dim repeatCount As Int32 = mIndentSize * mIndent
         If repeatcount > 0 Then
            sb.Append(" "c, mIndentSize * mIndent)
         End If
         Return sb.ToString
      End Function

      Public Function InsertNLIndent() As String
         Return Microsoft.VisualBasic.ControlChars.CrLf & InsertIndent()
      End Function
   End Class

End Class
